core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
-    //
-    // This selection of a peer and replication is continued in a loop until one of the
-    // following 3 conditions is fulfilled:
-    // (i) specified number of peers have been replicated to
-    // (ii) too many failures in replicating to peers
-    // (iii) no peer left to replicate to
-    //
-    while (!done) {
-      getRandomPeer() match {
-        case Some(peer) =>
-          try {
-            val onePeerStartTime = System.currentTimeMillis
-            logTrace(s"Trying to replicate $blockId of ${data.size} bytes to $peer")
-            blockTransferService.uploadBlockSync(
-              peer.host,
-              peer.port,
-              peer.executorId,
-              blockId,
-              new NettyManagedBuffer(data.toNetty),
-              tLevel,
-              classTag)
-            logTrace(s"Replicated $blockId of ${data.size} bytes to $peer in %s ms"
-              .format(System.currentTimeMillis - onePeerStartTime))
-            peersReplicatedTo += peer
-            peersForReplication -= peer
-            replicationFailed = false
-            if (peersReplicatedTo.size == numPeersToReplicateTo) {
-              done = true  // specified number of peers have been replicated to
+    val numPeersToReplicateTo = level.replication - 1
+
+    @tailrec def replicateBlock(
+      numFailures: Int,
+      peersForReplication: Seq[BlockManagerId],
+      peersReplicatedTo: Set[BlockManagerId],
+      peersFailedToReplicateTo: Set[BlockManagerId]): Set[BlockManagerId] = {
+
+      if (numFailures > maxReplicationFailures
+                || peersForReplication.isEmpty
+                || peersReplicatedTo.size == numPeersToReplicateTo) {
+        // This selection of a peer and replication is continued in a loop until one of the
+        // following 3 conditions is fulfilled:
+        // (i) specified number of peers have been replicated to
+        // (ii) too many failures in replicating to peers
+        // (iii) no peer left to replicate to
+        peersReplicatedTo
+      } else {
+        val peer = peersForReplication.head
+        val (updatedNumFailures, updatedPeers, updatedReplicatedPeers, updatedFailedPeers) = try {
+          val onePeerStartTime = System.currentTimeMillis
+          logTrace(s"Trying to replicate $blockId of ${data.size} bytes to $peer")
+          blockTransferService.uploadBlockSync(
+            peer.host,
+            peer.port,
+            peer.executorId,
+            blockId,
+            new NettyManagedBuffer(data.toNetty),
+            tLevel,
+            classTag)
+          logTrace(s"Replicated $blockId of ${data.size} bytes to $peer" +
+            s" in ${System.currentTimeMillis - onePeerStartTime} ms")
+          // the block was replicated, lets update state and move ahead
+          (numFailures,
+            peersForReplication.tail,
+            peersReplicatedTo + peer,
+            peersFailedToReplicateTo)
+        } catch {
+          case e: Exception =>
```

Comment:
The intended behavior here is to keep trying with other available options, if there is any exception here, till we run out of options or have a high number of failures. The method does a "best-effort" replication in that sense. It was the same behavior in the earlier implementation as well.